axlearn/common/attention.py

```diff
@@ -837,25 +837,41 @@ def extend_step(
         q_proj, k_proj, v_proj = self.forward(query, **kv_kwargs, query_positions=query_positions)
         updated_state = dict(time_step=time_step + num_query_steps)
         if kv_state is None:
-            # Update the cache via dynamic slice. [B, S, N, H].
+            # Update the cache via one-hot broadcast and addition.
             cached_key = cached_states["key"]
             cached_value = cached_states["value"]
 
-            # Ensure that we accumulate using the original dtype.
-            k_proj = k_proj.astype(cached_key.dtype)
-            v_proj = v_proj.astype(cached_value.dtype)
-
-            # TODO(dhwang2): jax.lax.dynamic_update_slice_in_dim is generally faster than advanced
-            # indexing, but an unusual slowdown was observed, with RLHF sampling taking up to
-            # 3 hours per run. Investigate and fix it.
-            # Note: All X_idx are small, so generating them on-demand is not costly.
-            b, _, n, h = cached_key.shape
-            b_idx = jnp.arange(b)[:, None, None, None]
-            t_idx = (jnp.arange(k_proj.shape[1])[None] + time_step[:, None])[:, :, None, None]
-            n_idx = jnp.arange(n)[None, None, :, None]
-            h_idx = jnp.arange(h)[None, None, None, :]
-            k_proj = cached_key.at[b_idx, t_idx, n_idx, h_idx].set(k_proj)
-            v_proj = cached_value.at[b_idx, t_idx, n_idx, h_idx].set(v_proj)
+            target_len = cached_key.shape[1]
+
+            # Create one-hot encodings for each position in the step range
+            # time_step is [B], need to create [B, step] indices
+            base_indices = time_step[:, None] + jnp.arange(num_query_steps)  # [B, step]
+            oh_indices = jax.nn.one_hot(
+                base_indices, target_len, dtype=k_proj.dtype
+            )  # [B, step, T]
+
+            # Reshape to [B, step, T, 1, 1] to broadcast across N and H dimensions
+            oh_indices = oh_indices[:, :, :, None, None]
+            negated_oh_indices = (1 - oh_indices).astype(cached_key.dtype)
+            negated_oh_indices = jnp.prod(negated_oh_indices, axis=1)  # [B, T, 1, 1]
+
+            # Reshape projections to align with one-hot encoding
+            # from [B, step, N, H] to [B, step, 1, N, H]
+            k_proj = k_proj[:, :, None, :, :]
+            v_proj = v_proj[:, :, None, :, :]
+
+            # Perform the update
+            # First multiply with one-hot encodings: [B, step, T, N, H]
+            k_proj = (k_proj * oh_indices).astype(cached_key.dtype)
+            v_proj = (v_proj * oh_indices).astype(cached_value.dtype)
+
+            # Sum across step dimension to get final projection values: [B, T, N, H]
+            k_proj = jnp.sum(k_proj, axis=1)
+            v_proj = jnp.sum(v_proj, axis=1)
+
+            # Combine with cached values
+            k_proj = (cached_key * negated_oh_indices) + k_proj
+            v_proj = (cached_value * negated_oh_indices) + v_proj
```

Comment:
I wonder if we can simplify a bit:
```suggestion
            source_len = cached_key.shape[1]
            
            # Create a dispatch matrix of shape [B, T=step, S].
            oh_indices = jax.nn.one_hot(
                time_step[:, None] + jnp.arange(num_query_steps), source_len, dtype=k_proj.dtype
            )
            # Create a mask of shape [B, S, 1, 1].
            negated_oh_indices = (1 - oh_indices.sum(axis=1))[..., None, None]
            k_proj = jnp.einsum("bt...,bts->bs...", k_proj, oh_indices)
            v_proj = jnp.einsum("bt...,bts->bs...", v_proj, oh_indices)
            k_proj = cached_key * negated_oh_indices + k_proj
            v_proj = cached_value * negated_oh_indices + v_proj
```

Reply:
This looks great! I updated the impl.

Reply:
Is this code as fast as `dynamic_update_slice_in_dim` during prefill?  
I'm a bit concerned about having to create a [B, T, S] tensor, especially for very long prefill like pdf or speech or video.
I want to take this opportunity to learn the speed relationship between **dynamic slice, broadcast, and advanced indexing**.

Reply:
Indeed, so I did a profile with 8k length prefill, dynamic slice and broadcast are roughly the same, and indexing is 2 orders of magnitude slower.